pkg/mutation/system_test.go

```diff
@@ -301,25 +301,193 @@ func TestMutation(t *testing.T) {
+	t.Run("mutate works on consistent state", func(t *testing.T) {
+		u := &unstructured.Unstructured{}
+		gotMutated, gotErr := s.Mutate(u, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "billing"}})
+		if !gotMutated {
+			t.Errorf("got Mutate() = %t, want true", gotMutated)
+		}
+		if gotErr != nil {
+			t.Fatalf("got Mutate() error = %v, want <nil>", gotErr)
+		}
+	})
+
+	// Put System in an inconsistent state.
+	err = s.Upsert(fooConflict)
+	if err == nil {
+		t.Fatalf("got Upsert() error = %v, want error", err)
+	}
+
+	// Since foo and foo-conflict define conflicting schemas, neither is executed.
+	t.Run("", func(t *testing.T) {
+		u2 := &unstructured.Unstructured{}
+		gotMutated, gotErr := s.Mutate(u2, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "billing"}})
+		// TODO(willbeason): Fix once System is updated to properly report conflicts (#1216).
+		if !gotMutated {
+			t.Errorf("got Mutate() = %t, want true", gotMutated)
+		}
+		// if gotMutated {
+		// 	t.Errorf("got Mutate() = %t, want false", gotMutated)
+		// }
+		if gotErr != nil {
+			t.Fatalf("got Mutate() error = %v, want <nil>", gotErr)
+		}
+	})
+
+	// Get the system back to a consistent state.
+	err = s.Remove(types.ID{Name: "foo-conflict"})
+	if err != nil {
+		t.Fatalf("got Remove() error = %v, want <nil>", err)
+	}
+
+	// Mutations are performed again.
+	t.Run("mutations performed after conflict removed", func(t *testing.T) {
+		u3 := &unstructured.Unstructured{}
+		gotMutated, gotErr := s.Mutate(u3, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "billing"}})
+		if !gotMutated {
+			t.Errorf("got Mutate() = %t, want true", gotMutated)
+		}
+		if gotErr != nil {
+			t.Fatalf("got Mutate() error = %v, want <nil>", gotErr)
+		}
+	})
+}
+
+func TestSystem_DontApplyConflictingMutationsRemoveOriginal(t *testing.T) {
+	// Two conflicting mutators.
+	foo := &fakeMutator{
+		MID:    types.ID{Name: "foo"},
+		MPath:  mustParse("spec.containers[name: foo].image"),
+		GVKs:   []schema.GroupVersionKind{{Version: "v1", Kind: "Pod"}},
+		Labels: map[string]string{"active": "true"},
+	}
+	fooConflict := &fakeMutator{
+		MID:    types.ID{Name: "foo-conflict"},
+		MPath:  mustParse("spec.containers.image"),
+		GVKs:   []schema.GroupVersionKind{{Version: "v1", Kind: "Pod"}},
+		Labels: map[string]string{"active": "true"},
+	}
+
+	// Put System in an inconsistent state.
+	s := NewSystem()
+	err := s.Upsert(foo)
+	if err != nil {
+		t.Fatalf("got Upsert() error = %v, want <nil>", err)
+	}
+	err = s.Upsert(fooConflict)
+	if err == nil {
+		t.Fatalf("got Upsert() error = %v, want error", err)
+	}
+	gotErr := s.Remove(types.ID{Name: "foo"})
+	if gotErr != nil {
+		t.Fatalf("got Remove() error = %v, want <nil>", gotErr)
+	}
+
+	u := &unstructured.Unstructured{}
+	gotMutated, gotErr := s.Mutate(u, &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: "billing"}})
+	// TODO(willbeason): Fix once System is updated to properly report conflicts (#1216).
+	if gotMutated {
```

Comment:
Aren't we expecting this to mutate since we are removing the original conflict?

Reply:
Done.

Reply:
Should we be testing against `!gotMutated` instead, since we are expecting a mutation?

Also, I think we can remove the commented out code below and the TODO above.

Reply:
The test logic is correct. We do not expect any mutations as the original Mutator was removed, so at this stage there are no more mutators. I opted not to fully explain this in test comments since it'll completely change in the next PR.

The error comment just wasn't updated. It now correctly states "want false".